src/bvals/boundary_conditions_generic.hpp

```diff
@@ -179,70 +184,74 @@ void GenericBC(std::shared_ptr<MeshBlockData<Real>> &rc, bool coarse,
-                  : (X2 ? IndexDomain::inner_x2 : IndexDomain::inner_x3))
-            : (X1 ? IndexDomain::outer_x1
-                  : (X2 ? IndexDomain::outer_x2 : IndexDomain::outer_x3));
+    constexpr IndexDomain domain =
+        INNER ? (X1 ? IndexDomain::inner_x1
+                    : (X2 ? IndexDomain::inner_x2 : IndexDomain::inner_x3))
+              : (X1 ? IndexDomain::outer_x1
+                    : (X2 ? IndexDomain::outer_x2 : IndexDomain::outer_x3));
 
-  // used for reflections
-  const int offset = 2 * ref + (INNER ? -1 : 1);
+    // used for reflections
+    const int offset = 2 * ref + (INNER ? -1 : 1);
 
-  // used for derivatives
-  const int offsetin = INNER;
-  const int offsetout = !INNER;
-  pmb->par_for_bndry(
-      PARTHENON_AUTO_LABEL, nb, domain, el, coarse,
-      KOKKOS_LAMBDA(const int &l, const int &k, const int &j, const int &i) {
-        if (TYPE == BCType::Reflect) {
-          const bool reflect = (q(b, el, l).vector_component == DIR);
-          q(b, el, l, k, j, i) =
-              (reflect ? -1.0 : 1.0) *
-              q(b, el, l, X3 ? offset - k : k, X2 ? offset - j : j, X1 ? offset - i : i);
-        } else if (TYPE == BCType::FixedFace) {
-          q(b, el, l, k, j, i) = 2.0 * val - q(b, el, l, X3 ? offset - k : k,
-                                               X2 ? offset - j : j, X1 ? offset - i : i);
-        } else if (TYPE == BCType::ConstantDeriv) {
-          Real dq = q(b, el, l, X3 ? ref + offsetin : k, X2 ? ref + offsetin : j,
-                      X1 ? ref + offsetin : i) -
-                    q(b, el, l, X3 ? ref - offsetout : k, X2 ? ref - offsetout : j,
-                      X1 ? ref - offsetout : i);
-          Real delta = 0.0;
-          if (X1) {
-            delta = i - ref;
-          } else if (X2) {
-            delta = j - ref;
+    // used for derivatives
+    const int offsetin = INNER;
+    const int offsetout = !INNER;
+    pmb->par_for_bndry(
+        PARTHENON_AUTO_LABEL, nb, domain, el, coarse, fine,
+        KOKKOS_LAMBDA(const int &l, const int &k, const int &j, const int &i) {
+          if (TYPE == BCType::Reflect) {
+            const bool reflect = (q(b, el, l).vector_component == DIR);
+            q(b, el, l, k, j, i) =
+                (reflect ? -1.0 : 1.0) * q(b, el, l, X3 ? offset - k : k,
+                                           X2 ? offset - j : j, X1 ? offset - i : i);
+          } else if (TYPE == BCType::FixedFace) {
+            q(b, el, l, k, j, i) =
+                2.0 * val - q(b, el, l, X3 ? offset - k : k, X2 ? offset - j : j,
+                              X1 ? offset - i : i);
+          } else if (TYPE == BCType::ConstantDeriv) {
+            Real dq = q(b, el, l, X3 ? ref + offsetin : k, X2 ? ref + offsetin : j,
+                        X1 ? ref + offsetin : i) -
+                      q(b, el, l, X3 ? ref - offsetout : k, X2 ? ref - offsetout : j,
+                        X1 ? ref - offsetout : i);
+            Real delta = 0.0;
```

Comment:
Given the complexity of this code (and above in this file) is it worth to update the new test case for use of non-periodic boundary conditions?

Reply:
yes, I think we want test coverage for all types of boundary conditions, different forest configurations, sparse and non-sparse, and regular and fine fields. Once the rest of these PRs that rely on this one are in, I was going to make another PR adding regression tests, add gold files, and maybe remove the old sparse advection test.